pkg/cache/v3/delta.go

```diff
@@ -0,0 +1,93 @@
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package cache
+
+import (
+	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
+	"github.com/envoyproxy/go-control-plane/pkg/log"
+	"github.com/envoyproxy/go-control-plane/pkg/server/stream/v3"
+)
+
+// Respond to a delta watch with the provided snapshot value. If the response is nil, there has been no state change.
+func respondDelta(request *DeltaRequest, value chan DeltaResponse, st *stream.StreamState, snapshot Snapshot, log log.Logger) *RawDeltaResponse {
+	resp, err := createDeltaResponse(request, st, snapshot)
+	if err != nil {
+		if log != nil {
+			log.Errorf("Error creating delta response: %v", err)
+		}
+		return nil
+	}
+
+	// Only send a response if there were changes
+	if len(resp.Resources) > 0 || len(resp.RemovedResources) > 0 {
+		if log != nil {
+			log.Debugf("node: %s, sending delta response with resources: %v removed resources %v wildcard: %t",
+				request.GetNode().GetId(), resp.Resources, resp.RemovedResources, st.IsWildcard)
+		}
+		value <- resp
+		return resp
+	}
+	return nil
+}
+
+func createDeltaResponse(request *DeltaRequest, st *stream.StreamState, snapshot Snapshot) (*RawDeltaResponse, error) {
+	resources := snapshot.GetResources((request.TypeUrl))
+
+	// variables to build our response with
+	nextVersionMap := make(map[string]string)
+	filtered := make([]types.Resource, 0, len(resources))
+	toRemove := make([]string, 0)
+
+	// Wildcard requests originate from CDS or LDS. If detect one, we want to track all resources as well as track them in the version map
+	if st.IsWildcard {
+		for name, r := range resources {
+			// Since we've already precomputed the version hashes of the new snapshot,
+			// we can just set it here to be used for comparison later
+			version := snapshot.GetVersionMap()[request.TypeUrl][name]
+			nextVersionMap[name] = version
+			prevVersion, found := st.ResourceVersions[name]
+			if !found || (prevVersion != nextVersionMap[name]) {
+				filtered = append(filtered, r)
+			}
+		}
+	} else {
+		// Reply only with the requested resources. Envoy may ask each resource
+		// individually in a separate stream. It is ok to reply with the same version
+		// on separate streams since requests do not share their response states.
+		for name, prevVersion := range st.ResourceVersions {
+			if r, ok := resources[name]; ok {
+				nextVersion := snapshot.GetVersionMap()[request.TypeUrl][name]
+				if prevVersion != nextVersion {
+					filtered = append(filtered, r)
+				}
+				nextVersionMap[name] = nextVersion
+			} else {
+				if prevVersion != "" {
+					toRemove = append(toRemove, name)
+				}
```

Comment:
will this ever be true now that we don't leave the deleted entries in there with empty string? maybe we can remove if check and just assume that it should be removed?

Reply:
Oh hmm that's actually a good point. That entry won't ever exist since we delete so that check would never pass. That check is mute so I'll remove.